Mouse drag

PaulBreugnot opened the pull request:
Hello Gama developers,

I am not sure if this should be included in the GAMA 1.9.1 branch, but the purpose of this pull request is to add a complete support for `mouse_drag` events in Gama.

This features has notably been used in the context of the [interactive Camisol model](https://github.com/PaulBreugnot/u2worm/tree/interface). Indeed, `mouse_drag` events are very useful in interactive models, to allow users to "drag and drop" agents for example. This feature is even required to allow users to move agents on a touchscreen, since `mouse_move` events do not exist on those devices (since the "mouse" is necessarily "down" while you move your finger on the screen).

However, implementing it required several sub-features, that should be approved:
- Dragging an agent has no sense when dragging causes the camera to move. This was not a problem for OpenGL contexts, where the camera can be locked. But in other cases, I implemented the ability to lock any `Zoomable` display, including Java2D displays. (see ccc2d0d).
- I noticed a bug about the `#user_location` management in OpenGL contexts: the #user_location was updated **after** the corresponding mouse events was triggered, what was critical on touchscreens were `mouse_down` events were systematically one event late for example. See the following discussion for more technical details on how the issue has been solved: https://groups.google.com/g/gama-platform/c/wqWxTsndKdg
- Adding `mouse_drag` event handling and GAMA constants everywhere it was necessary. This basically consisted in adding a line that corresponds to `mouse_drag` events everywhere `mouse_move` events was already handled, and should not cause any trouble.

I implemented those features a few month ago and did not had the time to perform the pull request until now, so I tried to carefully and regularly merge changed from the GAMA_1.9.0 into mine, without breaking anything. Let me now if you think there might have been merge issues.

Here is a table that sums up the configurations on which I tested the features: (:o: means untested)
| | Laptop | Touchscreen |
|-|-|-|
| Linux            | :heavy_check_mark: | :o: |
| MacOS         | :o: | :o: |
| Windows 10 | :heavy_check_mark: | :heavy_check_mark: |

The `Mouse Drag` example model of the `User Interaction` category can be used to easily test the feature.

I added Javadoc within the code for features I added, but I am not familiar with the Gama documentation generation.

Let me now if you need any information.

Paul

Commit:
Ability to lock any Zoomable display.

This commits adds the toggleLock() method to the Zoomable interface.
When a Zoomable item is locked, zoomIn(), zoomOut() and dragging of the
view pane have no effect, but zoomFit() is still active.

The ability to lock a display was already available for OpenGL displays,
thanks to the Camera interface. The feature has been generalised, and is
now available for:
- Java2DDisplaySurface
- GISFileViewer
- ImageViewer

The already existing lock control has been added to toolbars
accordingly, to any Zoomable display (see ZoomController).

Commit:
Implements mouse_drag events handling.

Adds a mouse_drag event to GAML. The mouse_drag event is the kind of events
emitted when the mouse is moved while the mouse button is down (while mouse_move
events are emitted when the mouse button is up). This addition adds significant
abilities to GAML, since mouse_move events (or any other kind of event) was not
sent while the mouse button was down (at least on linux platforms), what
prevented to implement "drag and drop" like features within GAMA models.

- ummisco.gama.annotations: The mouse_drag keyword has been added to GAML
- msi.gama.core: EventLayer, EventLayerStatement, IEventLayerListener and
  MouseEventLayerDelegate have been adapted to handle mouse dragged events.
- msi.gama.models: A Mouse Drag GAMA model has been added to the library models
  in the User Interaction section to demonstrate how the feature can be used to
  drag and drop an agent in both OpenGL and Java2D displays.
- ummisco.gama.ui.experiment: The LayeredDisplayMultiListener now properly
  handles drag events
- ummisco.gama.java2D: The Java2DDisplaySurface now dispatches SWT.DragDetect
  events
- ummisco.gama.opengl: The SWTOpenGLDisplaySurface now dispatches SWT.DragDetect
  events

- The feature is still not properly supported on tablets, but this hopefully
  will be fixed in next commits.

Commit:
Solves a mouse_down issue, critical on touch screens

In OpenGL displays, the mouse_down events was not properly handled since
the regular SWT event layer triggered the GAML mouse_down events, and
only after that (from an other SWT event) the OpenGL CameraHelper
updated the world position, i.e. the user_location. In consequence, the
user_location available from the mouse_down GAML action was not up to
date. This was not critical when using a mouse, since in this case
mouse_down events are always preceded by mouse_move events that also
update the user_location. However, this was critical on touch screens,
where independent mouse_down events are triggered from random places
without mouse_move events.

In order to solve the issue, the world position in OpenGL displays is
now updated from the SWTOpenGLDisplaySurface, from the same SWT mouse
down events that trigger corresponding GAML actions, and not from the
CameraHelper. However, the CameraHelper is still free to use the updated
world position, even if it also manages its own and independent
mousePosition.

Commit:
Fixes OpenGL issues introduced in the previous commit.

SWTOpenGLDisplaySurface::setMousePosition() is now based on
CameraHelper::getWorldPositionFrom() instead of
CameraHelper::computeMouseLocationInTheWorld(), as the
setMousePosition() is called from a non OpenGL context.

In consequence, the SWTOpenGLDisplaySurface now manages its own
world_position, and the CameraHelper::positionInTheWorld has gone
private for consistency, as it is just used internally.

Commit:
Merges branch 'GAMA_1.9.0' into mouse_drag

Commit:
Merge branch 'GAMA_1.9.0' into mouse_drag

AlexisDrogoul committed:
Update github-travis.yml

Remove cron scheduling

AlexisDrogoul committed:
Update github-daily-doc-wiki-refresh.yaml

Remove cron

Commit:
Merge branch 'GAMA_1.9.0' into mouse_drag

Commit:
Merge branch 'GAMA_1.9.1' into mouse_drag

AlexisDrogoul commented:
Super work @PaulBreugnot ! However,  it seems there is a conflict but I cant really figure what it is. And also: it would be nice to retarget this PR to https://github.com/gama-platform/gama/tree/GAMA_1.9.2. Thanks !

Commit:
Merge branch 'GAMA_1.9.2' into mouse_drag

PaulBreugnot commented:
Thanks for your answer!
I changed the base branch to GAMA_1.9.2, and fixed the conflict, that was only due to the Windows End Of Line characters that conflict with Unix ones...
I don't know what is the Gama versioning policy, but I see GAMA_1.9.2 is for GAMA_1.9.0 maintenance. The current PR is a bit more than maintenance since it adds some minor features. But I think it should be merged relatively quickly before changes become incompatible, so I'm not sure it can wait for GAMA_1.10. I hope it's not a problem.

Commit:
Merge branch 'GAMA_1.9.2' into mouse_drag

Adapt the "Mouse Drag.gaml" example to the new way of specifying event
actions, introduced in commit 04cf1e6.
